File: android/activitysources/src/main/java/com/fjuul/sdk/activitysources/entities/internal/healthconnect/sync_metadata/HCSyncMetadataStore.java
package com.fjuul.sdk.activitysources.entities.internal.healthconnect.sync_metadata;

import java.util.Date;
import java.util.Optional;

import com.fjuul.sdk.core.storage.Storage;

public class HCSyncMetadataStore {
    private static final String LAST_SYNC_TIME_KEY = "health_connect_last_sync_time";
    private static final String LOWER_DATE_BOUNDARY_KEY = "health_connect_lower_date_boundary";

    private final Storage storage;

    public HCSyncMetadataStore(@NonNull Storage storage) {
        this.storage = storage;
    }

    public void setLastSyncTime(@NonNull Date date) {
        storage.put(LAST_SYNC_TIME_KEY, date.getTime());
    }

    @NonNull
    public Optional<Date> getLastSyncTime() {
        Long timestamp = storage.getLong(LAST_SYNC_TIME_KEY);
        if (timestamp == null) {
            return Optional.empty();
        }
        return Optional.of(new Date(timestamp));
    }

    public void setLowerDateBoundary(@NonNull Date date) {
        storage.put(LOWER_DATE_BOUNDARY_KEY, date.getTime());
    }

    @NonNull
    public Optional<Date> getLowerDateBoundary() {
        Long timestamp = storage.getLong(LOWER_DATE_BOUNDARY_KEY);
        if (timestamp == null) {
            return Optional.empty();
        }
        return Optional.of(new Date(timestamp));
    }
} 

File: android/activitysources/src/main/java/com/fjuul/sdk/activitysources/entities/internal/healthconnect/sync_metadata/HCSyncMetadataStore.kt
package com.fjuul.sdk.activitysources.entities.internal.healthconnect.sync_metadata

import androidx.health.connect.client.records.Record
import com.fjuul.sdk.core.entities.IStorage
import java.time.Duration
import java.time.Instant

class HCSyncMetadataStore(private val storage: IStorage) {
    private val syncMetadata: MutableMap<String, Instant> = HashMap()

    fun isNeededToSync(recordType: Class<out Record?>): Boolean {
        val key = getKeyForRecordType(recordType)
        val lastSync = syncMetadata[key] ?: return true
        return Duration.between(lastSync, Instant.now()).compareTo(SYNC_INTERVAL) > 0
    }

    fun saveSyncMetadata(recordType: Class<out Record?>) {
        val key = getKeyForRecordType(recordType)
        syncMetadata[key] = Instant.now()
        storage[key] = Instant.now().toEpochMilli().toString()
    }

    private fun getKeyForRecordType(recordType: Class<out Record?>): String {
        return KEY_PREFIX + recordType.simpleName
    }

    companion object {
        private const val KEY_PREFIX = "health_connect_sync."
        private val SYNC_INTERVAL: Duration = Duration.ofHours(24)
    }
}
